### DISCUSSION TITLE
Aggregating ports with a virtual chassis

### DISCUSSION THREAD

**Discussion by jschewebbn** [netbox-community/netbox#12525] [Ideas]
The existing virtual chassis implementation shows all of the interfaces from all member devices on the master device. I'd like to be able to display all of the ports as well. The use case I have is multiple patch panels in a rack. I've modeled all of my patch panels separately as they are separate physical devices. The ports on all patch panels have unique labels as if they were all part of a single patch panel. When creating cable connections I usually don't care which specific patch panel device, I only care about the label. However other times I need to know the physical device. So I'd like to create a virtual chassis out of all of my patch panels and then be able to view all ports on the master device for when connecting cables. Is this a feature that the netbox community and maintainers would be willing to accept if I coded it up? Or is there another way to do this?

**Comment by mmfreitas** [comment]
I would find this quite useful since the patching in our data center are also separate devices which are mounted in a chassis, this would allow to list all front-ports and rear-ports on the virtual chassis instead of per device.

**Comment by candlerb** [comment]
Have you tried modelling the overall patch panel as a Device with Module Bays, and adding Modules with the frontports/rearports?

In the past I've done this with Device Bays and child Devices for [these](https://www.completeconnect.co.uk/mx-xpress-pre-terminated-modular-fibre/) modular patch panels, before Modules were available in Netbox. This worked fine, but I had to give each module its own name, and drill down to the module to see the connections.

Now that Modules are available in Netbox, then all the ports should be visible on the parent device, but I've not tested it. Try it and see.

**Reply by mmfreitas** [reply]
That's one way to do it, we have modular patch pannels too, but in our case we don't want to completely delete the module when it is removed as it may have useful information on its ports or special notes, so we treat it as a device that can be moved between chassis and are able to track this inventory. Having front ports and rear ports display on virtual chassis would be good enough to view it all in one place.

**Reply by candlerb** [reply]
Can you not move modules between devices then? I'm pretty sure you can with Inventory Items, but haven't tried with modules.

**Comment by jschewebbn** [comment]
The last time I looked at modules I need to dig into the device and then the module to see anything. That isn't a very useful option to me. If someone can demonstrate this on the demo instance, that would be great.

**Reply by candlerb** [reply]
All the frontports/rearports from all the modules are visible directly on the device:
![image](https://github.com/netbox-community/netbox/assets/44789/ac8bdf89-5c4e-4ea4-b414-f09fc12a1b77)

(at https://demo.netbox.dev/dcim/devices/128/front-ports/ until it gets wiped). That example has device "pp1" with 5 module bays, 2 of which are populated with a module with 12 frontports and 1 rearport (`F{module}-[1-12]` and `R{module}` respectively)

Also, if you examine any frontport, you can see which module it came from:
![image](https://github.com/netbox-community/netbox/assets/44789/57c40911-a4a1-4633-9950-382ec9bebab7)

However, modules *can't* be moved between devices. You have to delete the instance and recreate it somewhere else.  (But you *can* move it between module bays on the same device).

**Comment by jschewebbn** [comment]
OK, that is very interesting. Maybe that's the answer for the patch panels. I'll experiment with this more before I go write something new. I also see that interfaces show up now. My guess is that this is a change since I last upgraded netbox.

**Comment by jschewebbn** [comment]
I thought of an issue with using modules. I want to put my patch panels in my rack and they aren't always contiguous, there are switches and other hardware between them. I think that doesn't work with the modules option.

**Reply by candlerb** [reply]
Then clearly your patch panels are separate devices. Each device can still have its own set of modules; you just won't be able to move modules between devices, you'll have to delete and recreate.

You won't be able to see all the patching in one place either, but that's because it really isn't all in one place :-)

**Comment by jschewebbn** [comment]
That all makes sense. That's why I started looking at Virtual Chassis as this isn't something that is racked.

**Reply by candlerb** [reply]
Understood. In that case, the only option I can think of is to go to DCIM > Frontports, go to the Filters tab, and filter by Device Rack. It should show you all the frontports in that rack. If that works, you can then save that as a saved filter (and there's no need to use modules at all)

**Comment by jschewebbn** [comment]
However that doesn't work for my original use case of connecting a cable to one of the ports on the virtual patch panel.
I find it really nice to leverage this when connecting a cable to a group of switches in a virtual chassis. I select the master device and then have access to all of the interfaces. I'd like the same for front and rear ports.

**Reply by candlerb** [reply]
OK, I see. In Netbox 3.5 the cable connection has a new 'search' button next to the frontport selector, and you can search by rack, but you still have to drill down via the correct Device before you can select the target frontport.

**Reply by jschewebbn** [reply]
I've been working on a patch recently and something that appears to have changed is the ability to select the master device and see all interfaces when connecting a cable. I tried this on my own 3.6 instance and on the development site. I believe that previously I could select the master device and see all interfaces. Has anyone else noticed this? Is this intentional or an oversight in the recent change for how devices are selected?

**Reply by candlerb** [reply]
When you say "master device", are you still trying to combine multiple patch panels with frontports/rearports into a virtual chassis?? I don't know if that's ever worked or been supported.

If these are active devices with *interfaces* rather than *frontports* then I suggest you raise an issue with specific reproduction steps. I think you should be able to see all interfaces on the master device. If you can see them in the interfaces table but not in the filter when connecting a cable, then that seems like an inconsistency to me.

If you are able to reproduce this with devices+interfaces, and it gets fixed, the fix *might* also make it work for devices+frontports (but no guarantees).

**Reply by jschewebbn** [reply]
I'm going back and testing the functionality that I believe worked before so that I can recreate it with ports. So I'm connecting a cable to to a device and selecting the master device and expecting to see all interfaces across all devices in the virtual chassis.

Visiting the master device itself shows all interfaces across all devices in the virtual chassis.

**Reply by jschewebbn** [reply]
https://github.com/netbox-community/netbox/issues/14336 submitted as I'm not sure if this is a feature request or a bug.

**Comment by jschewebbn** [comment]
My ticket was moved to a discussion over at https://github.com/netbox-community/netbox/discussions/14354

Thinking about this more and looking at how the cable connections changed in 3.6, I'm planning on making virtual chassis a more real device. I'm going to add the ability to select them as a device in the cable connection dialog and to show the interfaces, front ports and rear ports of all member devices.

**Comment by jschewebbn** [comment]
PR for adding ports support is in at https://github.com/netbox-community/netbox/pull/14783

**Comment by jschewebbn** [comment]
Trying to put in a feature request for this change again at https://github.com/netbox-community/netbox/issues/14785
